sys/contrib/ck/include/ck_stack.h

```diff
@@ -39,7 +39,12 @@ typedef struct ck_stack_entry ck_stack_entry_t;
 
 struct ck_stack {
 	struct ck_stack_entry *head;
+#ifdef __CHERI_PURE_CAPABILITY__
+	/* Can not reduce alignment on pointers */
+	char *generation;
+#else
 	char *generation CK_CC_PACKED;
```

Comment:
Why is this even there in the first place? It doesnt make any sense to me.

Reply:
This seems to come from here https://github.com/concurrencykit/ck/commit/66ebe1a874e18d978776cd3238ee3c422774475f
But it is unclear why it was added as it makes no sense to me either.

Reply:
Upon further inspection it seems that the generation pointer is actually a counter. Very weird choice of typing. I think it should become a `size_t` counter.

Reply:
The atomics plumbing in CK is a bit sad. They have not atomic operations defined on `long`, only explicit `*_64` variants and it looks like they defined it to be a pointer so that they could use `ck_pr_cas_ptr_2_value()` and take advantage of 2-register atomic load/stores in some architectures. This particular code path also seems unused, we don't seem to have a `ck_pr_cas_ptr_2_value()` implementation for mips and riscv, so I'm not sure why it does not fail at build time.
As an aside, the ARM morello implementation will fail on this as well. As an aside it's quite annoying to have to maintain 2 sets of atomics for the kernel in machine and ck separately.

Reply:
The CK_CC_PACKED isn't going to be doing anything here other than throw up a compiler warning for us, surely? The `head` pointer will enforce alignment of the struct, itself and, since it's immediately afterwards, `generation`.

Reply:
I feel like we should just not define CK_CC_PACKED in sys/contrib/ck/include/gcc/ck_cc.h when in purecap mode causing it do be defined to nothing later in sys/contrib/ck/include/ck_cc.h.

Reply:
That sounds like a recipe for things becoming unpacked that should be packed, though I guess in practice there should never be anything like that in CK... CK_CC_PACKED should just die upstream

Reply:
The only place where that is being used is for the generation counter in the ABA-safe pointer patterns.
AFAIK the rationale for it being there is that as for the C standard the compiler is allowed to insert arbitrary padding and that could somehow break the cas_2value primitive that they are using. I don't have enough insight about the C standard to discount this but it feels wrong, I'm not sure it should matter if there is padding as long as the two pointer-sized values are contiguous and aligned.

Reply:
Surely you can just static_assert(sizeof(x) == 16, "") to see there is no padding?